MathlibTest/FieldSimp.lean

```diff
@@ -359,30 +359,107 @@ end
 
 /-! ## Cancel denominators from equalities -/
 
-/-! ### Most common use case: Cancel denominators to something solvable by `ring`
+/-! ### Finishing tactic
 
-When (eventually) this is robust enough, there should be a `field` tactic
+The `field` tactic is a finishing tactic for equalities in fields.
+Effectively it runs `field_simp` to clear denominators, then hands the result to `ring1`.
 -/
 
-macro "field" : tactic => `(tactic | (try field_simp) <;> ring1)
-
 example : (1:ℚ) / 3 + 1 / 6 = 1 / 2 := by field
 example {x : ℚ} (hx : x ≠ 0) : x * x⁻¹ = 1 := by field
 example {a b : ℚ} (h : b ≠ 0) : a / b + 2 * a / b + (-a) / b + (- (2 * a)) / b = 0 := by field
+
+-- example from the `field` docstring
 example {x y : ℚ} (hx : x + y ≠ 0) : x / (x + y) + y / (x + y) = 1 := by field
 
 example {x y : ℚ} (hx : 0 < x) :
     ((x ^ 2 - y ^ 2) / (x ^ 2 + y ^ 2)) ^ 2 + (2 * x * y / (x ^ 2 + y ^ 2)) ^ 2 = 1 := by
   field
 
--- example from the `field_simp` docstring
 example {K : Type*} [Field K] (a b c d x y : K) (hx : x ≠ 0) (hy : y ≠ 0) :
     a + b / x + c / x ^ 2 + d / x ^ 3 = a + x⁻¹ * (y * b / y + (d / x + c) / x) := by
   field
 
+-- example from the `field` docstring
 example {a b : ℚ} (ha : a ≠ 0) : a / (a * b) - 1 / b = 0 := by field
+
 example {x : ℚ} : x ^ 2 * x⁻¹ = x := by field
 
+-- example from `field` docstring
+example {K : Type*} [Field K] (hK : ∀ x : K, x ^ 2 + 1 ≠ 0) (x : K) :
+    1 / (x ^ 2 + 1) + x ^ 2 / (x ^ 2 + 1) = 1 := by
+  field [hK]
+
+-- `field` will suggest `field_simp` on failure, if `field_simp` does anything
+/--
+info: Try this: field_simp
+---
+error: unsolved goals
+x y z : ℚ
+hx : x + y ≠ 0
+⊢ 1 = z
+-/
+#guard_msgs in
+example {x y z : ℚ} (hx : x + y ≠ 0) : x / (x + y) + y / (x + y) = z := by field
+
+-- If `field` fails but `field_simp` also fails, we just throw an error
+/--
+error: ring failed, ring expressions not equal
+x y z : ℚ
+⊢ x + y = z
+-/
+#guard_msgs in
+example {x y z : ℚ} : x + y = z := by field
+
+/-
+The `field` tactic differs slightly from `field_simp; ring1` in that it clears denominators only at
+the top level, not recursively in subexpressions.
+
+(`ring1` acts only at the top level, so for consistency we also clear denominators only at the top
+level.) -/
+/--
+info: Try this: field_simp
+---
+error: unsolved goals
+a b : ℚ
+f : ℚ → ℚ
+⊢ f (a * b) * (1 - 1) = 0
+-/
+#guard_msgs in
+example (a b : ℚ) (f : ℚ → ℚ) : f (a ^ 2 * b / a) - f (b ^ 2 * a / b) = 0 := by field
+
+-- (Compare with the example above: this is out of scope for `field`.)
+example (a b : ℚ) (f : ℚ → ℚ) : f (a ^ 2 * b / a) - f (b ^ 2 * a / b) = 0 := by
+  field_simp
+  ring1
+
+-- from `Analysis.Complex.UpperHalfPlane.MoebiusAction`
+-- `field` does not fully clear denominators in this example, but calling different normalizations
+-- in succession eventually succeeds
+-- (This example is used in the `field` docstring.)
```

Comment:
The overall paragraph refers to the next examples, but this line only to the next one, right? Can you move the blank line to before this line then?